instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py

```diff
@@ -0,0 +1,375 @@
+        SpanAttributes.HTTP_URL: _url,
+    }
+    return span_attributes
+
+
+def _prepare_headers(headers: typing.Optional[Headers]) -> httpx.Headers:
+    return httpx.Headers(headers)
+
+
+class SyncOpenTelemetryTransport(httpcore.SyncHTTPTransport):
+    """Sync transport class that will trace all requests made with a client.
+
+    Args:
+        transport: SyncHTTPTransport instance to wrap
+        tracer_provider: Tracer provider to use
+        request_hook: A hook that receives the span and request that is called
+            right after the span is created
+        response_hook: A hook that receives the span, request, and response
+            that is called right before the span ends
+    """
+
+    def __init__(
+        self,
+        transport: httpcore.SyncHTTPTransport,
+        tracer_provider: typing.Optional[TracerProvider] = None,
+        request_hook: typing.Optional[RequestHook] = None,
+        response_hook: typing.Optional[ResponseHook] = None,
+    ):
+        self._transport = transport
+        self._tracer = get_tracer(
+            __name__,
+            instrumenting_library_version=__version__,
+            tracer_provider=tracer_provider,
+        )
+        self._request_hook = request_hook
+        self._response_hook = response_hook
+
+    def request(
+        self,
+        method: bytes,
+        url: URL,
+        headers: typing.Optional[Headers] = None,
+        stream: typing.Optional[httpcore.SyncByteStream] = None,
+        ext: typing.Optional[dict] = None,
+    ) -> typing.Tuple[int, "Headers", httpcore.SyncByteStream, dict]:
+        """Add request info to span."""
+        if context.get_value("suppress_instrumentation"):
+            return self._transport.request(
+                method, url, headers=headers, stream=stream, ext=ext
+            )
+
+        span_attributes = _prepare_attributes(method, url)
+        _headers = _prepare_headers(headers)
+        span_name = _get_default_span_name(
+            span_attributes[SpanAttributes.HTTP_METHOD]
+        )
+        request = httpx.Request(method, url, headers=headers, stream=stream)
+
+        with self._tracer.start_as_current_span(
+            span_name, kind=SpanKind.CLIENT, attributes=span_attributes
+        ) as span:
+            if self._request_hook is not None:
+                self._request_hook(span, request)
+
+            inject(_headers)
+
+            status_code, headers, stream, extensions = self._transport.request(
+                method, url, headers=_headers.raw, stream=stream, ext=ext
+            )
+
+            _apply_status_code(span, status_code)
+
+            if self._response_hook is not None:
+                response = httpx.Response(
+                    status_code=status_code,
+                    headers=headers,
+                    stream=stream,
+                    ext=extensions,
+                    request=request,
+                )
+                response.read()
```

Comment:
Doesn't this change behavior of the library when response_hook is present? Does `response.read()` read the stream and move cursor. Wouldn't that result in surprising behavior for clients?

Generally I don't think it's worth it for instrumentations to do things like this just to satisfy response_hook. I think we should just pass all four returned values `status_code, headers, stream, extensions` to the resposne hook instead of creating a nice httpx.Response for it.

Reply:
The hooks are expected to accept library specific params and it doesn't always have to be a request or response object.

Reply:
I was looking at it from my own perspective when using it. I would want to deal with the Request and Response objects. I'll dig in further. I'll look at adding streaming requests and maybe some real requests to the tests and verify things work correctly.

Reply:
Yes, I can see it is desirable to work with response objects but instrumentation should only wrap things. It should be invisible and not try improve on interfaces provider by the library especially if it comes at a performance or behavioral change cost.

Reply:
Alright, so to make sure I understand correctly, the request hook should receive the tuple of raw information (method, url, headers, stream, extensions) like the request function parameters and the response hook should receive a tuple of (status_code, headers, stream, extensions) like in the return value? Any handling of the raw information (string URLs, headers as a dict, etc.) should be left to the user?

If that is the case, then I can do that, but I should also provide some common examples to make the hooks a little bit easier to use. Those values are tuned for the low-level API that people may not normally expect when using `httpx`, I assume, and could make it more difficult than it needs to be to use the hooks initially.

Reply:
Sorry, that's not what I meant. I meant if the function we are wrapping is returning raw values and we have to read some buffer just to be able to call a hook with nicer values then probably we should just call the hook with the raw values. Generally speaking an instrumentation should only observe what a library does and never affect how the library works. Reading the buffer/stream may seek the pointer to end resulting in unexpected behavior elsewhere. If that is not the case then it is a lot more acceptable but still for performance reasons, instrumentation should probably stay away from reading streams and wrapping objects just to make hooks look nicer. WDYT?

Reply:
Yeah, I understand wanting to avoid unexpected behavior. Back to the raw values now.